core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProcessBindingService.java

```diff
@@ -0,0 +1,116 @@
+import com.expediagroup.streamplatform.streamregistry.core.views.ProducerView;
+import com.expediagroup.streamplatform.streamregistry.model.ProcessBinding;
+import com.expediagroup.streamplatform.streamregistry.model.Status;
+import com.expediagroup.streamplatform.streamregistry.model.keys.ProcessBindingKey;
+import com.expediagroup.streamplatform.streamregistry.repository.ProcessBindingRepository;
+
+@Component
+@RequiredArgsConstructor
+public class ProcessBindingService {
+  private final ProcessBindingView processBindingView;
+  private final HandlerService handlerService;
+  private final ProcessBindingValidator processBindingValidator;
+  private final ProcessBindingRepository processBindingRepository;
+  private final ConsumerView consumerView;
+  private final ConsumerService consumerService;
+  private final ProducerView producerView;
+  private final ProducerService producerService;
+
+  @PreAuthorize("hasPermission(#processBinding, 'CREATE')")
+  public Optional<ProcessBinding> create(ProcessBinding processBinding) throws ValidationException {
+    if (processBindingView.get(processBinding.getKey()).isPresent()) {
+      throw new ValidationException("Can't create " + processBinding.getKey() + " because it already exists");
+    }
+    processBindingValidator.validateForCreate(processBinding);
+    processBinding.setSpecification(handlerService.handleInsert(processBinding));
+    return save(processBinding);
+  }
+
+  @PreAuthorize("hasPermission(#processBinding, 'UPDATE')")
+  public Optional<ProcessBinding> update(ProcessBinding processBinding) throws ValidationException {
+    val existing = processBindingView.get(processBinding.getKey());
+    if (!existing.isPresent()) {
+      throw new ValidationException("Can't update " + processBinding.getKey() + " because it doesn't exist");
+    }
+    processBindingValidator.validateForUpdate(processBinding, existing.get());
+    processBinding.setSpecification(handlerService.handleUpdate(processBinding, existing.get()));
+    return save(processBinding);
+  }
+
+  @PreAuthorize("hasPermission(#processBinding, 'UPDATE_STATUS')")
+  public Optional<ProcessBinding> updateStatus(ProcessBinding processBinding, Status status) {
+    processBinding.setStatus(status);
+    return save(processBinding);
+  }
+
+  private Optional<ProcessBinding> save(ProcessBinding processBinding) {
+    return Optional.ofNullable(processBindingRepository.save(processBinding));
+  }
+
+  @PostAuthorize("returnObject.isPresent() ? hasPermission(returnObject, 'READ') : true")
+  public Optional<ProcessBinding> get(ProcessBindingKey key) {
+    return processBindingView.get(key);
+  }
+
+  @PostFilter("hasPermission(filterObject, 'READ')")
+  public List<ProcessBinding> findAll(Predicate<ProcessBinding> filter) {
+    return processBindingView.findAll(filter).collect(toList());
+  }
+
+  @PreAuthorize("hasPermission(#processBinding, 'DELETE')")
+  public void delete(ProcessBinding processBinding) {
```

Comment:
The Process/ProcessBinding is like a specification of a small application with input Streams and output Streams. The ConsumerBindings and ProducerBindings referenced in the ProcessBinding are created specifically so that the application described by the Process can function. So, if the Process is being deleted all of the ConsumerBindings/ProducerBindings are no long relevant and should be deleted too.

We delete the Consumer and Producer here instead of the Bindings, because the handlers for Consumer and Producer should handle deleting the relevant bindings.

Reply:
So if the cascading is required, I'm still wondering if it's better conceptually applied on the `Process` rather than the `ProcessBinding`. For example, if we delete a `ProcessBinding`, but not the `Process`, does that mean we are left with a `Process` that references `Producer`/`Consumer`s that have been deleted?

Reply:
The Process doesn't have references to actual Consumers or Producers. The Process has, what amounts to a StreamKey as input and output references. During the reconciliation these StreamKey references are used to create Consumers/Producers, and their associated Bindings. There aren't actually any references to the Consumers and Producers, only references to their Bindings on the Stream Binding.

Since there are no direct references to Consumers/Producers or their bindings on the Process, we can't actually do any cascading from there. The Process Delete deletes itself and all of its bindings, which in turn deletes the Consumers/Producers and their bindings because that is where the references can be found.

Reply:
Cheers man, this this just made it click in my head how this fits together and how the reconciliation is actually gonna work for the new `Process` 😅 .

I guess a follow up question, should we be allowing the deletion of `Stream`s that are being used in `Process`es? If a Stream is deleted, it'll still be referenced from a Process. Any consumer/producer referenced in the ProcessBinding will have been deleted.

Reply:
@jamespfaulkner - this is a really good point, but I think that this is already protected in current deletion rules for streams. I'll check again, but I think that streams are already protected from deletion if the deletion request user does not have permissions to delete all of the consumers of the stream. Processes create consumers which will have security roles given to it from the process. If a user tries to delete a stream that has a process consumer referencing it, they will not be able to delete that stream if the user does not have permissions to delete the process consumer.

This does not prevent a user, who owns both the stream and the process, from deleting a stream from underneath an existing process. I think this logic may need to be thought out a bit more, and maybe an update may need to be made in subsequent change.

Reply:
Good point about the ownership. 

I was more thinking from an admin point of view. With mutation access to everything, we don't want to have to worry if a Stream is used in a Process when actioning deletes.

But yeah, this can go into a future PR. We're not going to have a Process being created by people outside our team for a  little while.

Reply:
Having actual Consumer and Producer lists as inputs and outputs in the Process makes a sort of symmetrical sense, since the CB and PB are in the ProcessBinding. But practically speaking, a full Consumer or Producer as an input or output seems a bit bloated from the input perspective. There is a lot of information in the Consumer and Producer that may not necessarily be required. Things like tags, description, security (can be inherited from Process) are not needed. We don't even need the entire Consumer/Producer key since the name of the Consumer/Producer should just be the Process name. The only information we need is the StreamKey, and for your example, a type that we wish to consume/produce. So it seems more light way (and to the point) to keep the Inputs/Outputs with just the StreamKey and an additional `type` field when we needed. Also, we just removed `locality` from the input, but where would this fit into a Consumer? It would have to go down into Configuration?

Reply:
I'm not sure. The ConsumerBinding and ProducerBinding have information that is useful to us about the physical consumer/producer, things like the certificate path, the type of consumer/producer connection and any metadata that is needed for that connection. This is opposed to the Consumer/Producer that really have no valuable information, everything about them is synthesized except for the streamKey and type.

This being the case I feel like it is valuable to have that direct link to the CB/PB, but not necessarily useful to have direct links to C/P. To be honest, the C/P are not really necessary at all in this context. Do they even need to exist for Process consumers and producers?